src/main/java/bgu/spl/net/impl/stomp/LibraryProtocol.java

```diff
@@ -13,59 +13,168 @@
-        if(receiptId != null) {
-            newFrame.addHeader("receiptId", "message-"+receiptId);
+        if (receiptId != null) {
+            Frame newFrame = new Frame("RECEIPT");
+            newFrame.addHeader("receipt-id", "message-" + receiptId);
+            _connections.send(_connectionId, newFrame);
         }
-        return newFrame;
+    }
+
+    private Frame createErrorFrame(String message) {
+        Frame f = new Frame("ERROR");
+        f.addHeader("message", message);
+        return f;
+    }
+
+    /**
+     * Creates a message frame. Note that this doesn't add the subscription-id, which will be added by the
+     * Connections interface once the message gets sent
+     *
+     * @param msg The message received by the client which triggered the call to this function
+     * @return The message to be sent to the users in the channel (without the subscription-id)
+     */
+    private Frame createMessageFrame(Frame msg) {
+        Frame f = new Frame("MESSAGE");
+        f.addHeader("Message-id", _messagesSent.toString());
+        _messagesSent++;
+        f.addHeader("destination", msg.getHeader("destination"));
+        f.addBody(msg.getBody());
+        return f;
     }
 
     private void handleSend(Frame msg) {
-        // TODO: Complete
+        if (!checkForHeader(msg, "destination")) return;
+        _connections.send(msg.getHeader("destination"), createMessageFrame(msg));
+        sendReceipt(msg);
     }
 
-    private Frame handleConnect(Frame msg) {
+    private void addErrorBody(Frame msg, Frame errorFrame, String info) {
+        errorFrame.addBody("The message:\n-----" + msg.toString() + "\n-----\n" + info);
+    }
+
+    /**
+     * Disconnect and send the frame received as a parameter
+     *
+     * @param f The frame to be sent
+     */
+    private void disconnectWithFrame(Frame f) {
+        _connections.send(_connectionId, f);
+        _connections.disconnect(_connectionId);
+        _shouldTerminate = true;
+    }
+
+    private void handleConnect(Frame msg) {
         Frame f = new Frame();
-        if ((_client = clientsByLogin.get(msg.getHeader("login"))) == null) {
+        if ((_client = clientsByLogin.get(msg.getHeader("login"))) == null) { // Is the user new?
             _client = new Client(msg.getHeader("host"), msg.getHeader("login"), msg.getHeader("passcode"));
             clientsByLogin.put(_client.login, _client);
-            f = createConnectedFrame();
-        } else if (_client.passcode.equals(msg.getHeader("passcode"))) {
-            f = createConnectedFrame();
-        } else {
+        } else if (_client.connected) { // Is the user already connected?
+            f.setCommand("ERROR");
+            f.addHeader("message", "User already logged in");
+            String receiptId = null;
+            if ((receiptId = msg.getHeader("receipt")) != null) f.addHeader("receipt-id", receiptId);
+            f.addBody("User already logged in");
+            disconnectWithFrame(f);
```

Comment:
why do you disconnect a connected user?

Reply:
This doesn't disconnect the user, this terminates the connection to the client that tried to log in to a user that is already logged in